pkg/component/runtime/manager.go

```diff
@@ -0,0 +1,569 @@
+package runtime
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/gofrs/uuid"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/elastic/elastic-agent-client/v7/pkg/client"
+	"github.com/elastic/elastic-agent-client/v7/pkg/proto"
+	"github.com/elastic/elastic-agent-libs/atomic"
+	"go.elastic.co/apm"
+	"go.elastic.co/apm/module/apmgrpc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/status"
+
+	"github.com/elastic/elastic-agent/internal/pkg/core/authority"
+	"github.com/elastic/elastic-agent/pkg/component"
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+)
+
+const (
+	// initialCheckinTimeout is the maximum amount of wait time from initial check-in stream to
+	// getting the first check-in observed state.
+	initialCheckinTimeout = 5 * time.Second
+	// maxCheckinMisses is the maximum number of check-in misses a component can miss before it is killed
+	// and restarted.
+	maxCheckinMisses = 3
+)
+
+var (
+	// ErrNoUnit returned when manager is not controlling this unit.
+	ErrNoUnit = errors.New("no unit under control of this manager")
+)
+
+// Manager for the entire runtime of operating components.
+type Manager struct {
+	proto.UnimplementedElasticAgentServer
+
+	logger     *logger.Logger
+	ca         *authority.CertificateAuthority
+	listenAddr string
+	tracer     *apm.Tracer
+
+	netMx    sync.RWMutex
+	listener net.Listener
+	server   *grpc.Server
+
+	waitMx    sync.RWMutex
+	waitReady map[string]waitForReady
+
+	mx      sync.RWMutex
+	current map[string]*componentRuntimeState
+
+	subMx         sync.RWMutex
+	subscriptions map[string][]*Subscription
```

Comment:
i dont see a need for generics here, we use just use channel from subscription which is exposed, we can hide this behind interface rather than having `*Subscription` reference though